mllib/src/test/scala/org/apache/spark/mllib/clustering/PowerIterationClusteringSuite.scala

```diff
@@ -0,0 +1,317 @@
+    val (ccenters, estCollected) = PIC.run(sc, G, nClusters, nIterations)
+    logger.info(s"Cluster centers: ${ccenters.mkString(",")} " +
+      s"\nEstimates: ${estCollected.mkString("[", ",", "]")}")
+    assert(ccenters.size == circleSpecs.length, "Did not get correct number of centers")
+
+  }
+
+}
+
+object PowerIterationClusteringSuite {
+  val logger = Logger.getLogger(getClass.getName)
+  val A = Array
+  val PIC = PowerIterationClustering
+
+  // Default sigma for Gaussian Distance calculations
+  val defaultSigma = 1.0
+
+  // Default minimum affinity between points - lower than this it is considered
+  // zero and no edge will be created
+  val defaultMinAffinity = 1e-11
+
+  def pdoub(d: Double) = f"$d%1.6f"
+
+  case class Point(label: Long, x: Double, y: Double) {
+    def this(x: Double, y: Double) = this(-1L, x, y)
+
+    override def toString() = s"($label, (${pdoub(x)},${pdoub(y)}))"
+  }
+
+  object Point {
+    def apply(x: Double, y: Double) = new Point(-1L, x, y)
+  }
+
+  case class CircleSpec(center: Point, radius: Double, noiseToRadiusRatio: Double,
+                        nPoints: Int, uniformDistOnCircle: Boolean = true)
+
+  def createConcentricCirclesData(circleSpecs: Seq[CircleSpec]) = {
+    import org.apache.spark.mllib.random.StandardNormalGenerator
+    val normalGen = new StandardNormalGenerator
+    var idStart = 0
+    val circles = for (csp <- circleSpecs) yield {
+      idStart += 1000
+      val circlePoints = for (thetax <- 0 until csp.nPoints) yield {
+        val theta = thetax * 2 * Math.PI / csp.nPoints
+        val (x, y) = (csp.radius * Math.cos(theta)
+          * (1 + normalGen.nextValue * csp.noiseToRadiusRatio),
+          csp.radius * Math.sin(theta) * (1 + normalGen.nextValue * csp.noiseToRadiusRatio))
+        (Point(idStart + thetax, x, y))
+      }
+      circlePoints
+    }
+    val points = circles.flatten.sortBy(_.label)
+    logger.info(printPoints(points))
+    points
+  }
+
+  def printPoints(points: Seq[Point]) = {
+    points.mkString("[", " , ", "]")
+  }
+
+  private[mllib] def printVector(dvect: BDV[Double]) = {
+    dvect.toArray.mkString(",")
+  }
+
+
+  /**
+   *
+   * Create an affinity matrix
+   *
+   * @param sc  Spark Context
+   * @param points  Input Points in format of [(VertexId,(x,y)]
+   *                where VertexId is a Long
+   * @param sigma   Sigma for Gaussian distribution calculation according to
+   *                [1/2 *sqrt(pi*sigma)] exp (- (x-y)**2 / 2sigma**2
+   * @param minAffinity  Minimum Affinity between two Points in the input dataset: below
+   *                     this threshold the affinity will be considered "close to" zero and
+   *                     no Edge will be created between those Points in the sparse matrix
+   * @return Tuple of (Seq[(Cluster Id,Cluster Center)],
+   *         Seq[(VertexId, ClusterID Membership)]
+   */
+  def createGaussianAffinityMatrix(sc: SparkContext,
```

Comment:
Let's do this in a follow-up PR. For unit tests, we can use a graph with precomputed distances and test against the expected results.